packages/react-components/src/components/VideoGallery/useVideoStreamLifecycleMaintainer.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import React, { useEffect, useRef } from 'react';
+import { VideoStreamOptions, CreateVideoStreamViewResult, ScalingMode } from '../../types';
+import { Cancellable, useCancellableTask } from '../utils/useCancellableTask';
+
+/**
+ * Helper hook to maintain the video stream lifecycle. This calls onCreateStreamView and onDisposeStreamView
+ * appropriately based on react lifecycle events and prop changes.
+ * This also handles calls to view.update* appropriately such as view.updateScalingMode().
+ *
+ * @private
+ */
+export const useVideoStreamLifecycleMaintainer = (props: {
+  isStreamAvailable?: boolean;
+  renderElementExists?: boolean;
+  isMirrored?: boolean;
+  scalingMode?: ScalingMode;
+  onCreateStreamView?: (options?: VideoStreamOptions) => Promise<void | CreateVideoStreamViewResult>;
+  onDisposeStreamView?: () => void;
+}): void => {
+  const { onCreateStreamView, onDisposeStreamView, isStreamAvailable, renderElementExists, isMirrored, scalingMode } =
+    props;
+
+  // HANDLING CHANGES TO VIDEO VIEW OPTIONS
+  //
+  // When VideoViewOptions change two things may happen:
+  //
+  // 1. Just the scaling mode has changed and we have access to the original StreamRenderView
+  //   - In this case we just need to call updateScalingMode on the view and do not need to recreate the stream view
+  // 2. isMirrored has changed, or scalingMode change and we don't have access to the original StreamRenderView
+  //   - In this case we need to dispose the old view and create a new one
+  //
+  // For scenario 1, we hold onto a ref to the scalingMode that persists across renders. When VideoViewOptions.scalingMode
+  // differs from the persistent ref, we know to call updateScalingMode. We then subsequently update the ref to the new value for
+  // future renders.
+  //
+  // For scenario 2, we extract the VideoViewOptions.isMirrored to ensure any change to the isMirrored prop triggers the useEffect
+  // that recreates the stream view. We must also here add an extra check for when the scaling mode changes and we do not have
+  // access to the original StreamView (and hence cannot call updateScalingMode). When this happens we must also trigger the useEffect
+  // that recreates the stream view.
+
+  const [streamRendererResult, setStreamRendererResult] = React.useState<CreateVideoStreamViewResult>();
+  const scalingModeRef = useRef(scalingMode);
+  const hasScalingModeChanged = scalingModeRef.current !== scalingMode;
+  const updatingScalingModeDirectly = hasScalingModeChanged && !!streamRendererResult;
+
+  const [triggerRescale, cancelRescale] = useCancellableTask();
+  const [triggerCreateStreamView, cancelCreateStreamView] = useCancellableTask();
+
+  if (isStreamAvailable && renderElementExists && scalingMode && updatingScalingModeDirectly) {
```

Comment:
Do we need to do useeffect here?

Reply:
For this it's simply redundant if we do because we have the `updatingScalingModeDirectly` check - so we could have it be we left it out because it just adds a few more lines of nesting and little bit extra complexity